### PR TITLE
🚧Refactor moq code to nsubstitute

### LINKED ISSUE #1006: 🚧Refactor moq code to nsubstitute
### Complete The Item Below

- [X] I have updated the title without removing the 🚧 emoji.

### Description

Refactor the following test files from **_Moq_** code to **_NSubstitute_** code

1. TextureFactoryTests.cs
1. TextureTests.cs
1. GpuData TypeExtensionsTests.cs
1. NativelnputFactoryTests.cs
1. ImageLoaderTests.cs
1. RenderMediatorTests.cs
1. FontRendererTests.cs
1. LineRendererTests.cs
1. ShapeRendererTests.cs
1. TextureRendererTests.cs
1. SystemDisplayServiceTests.cs

### Acceptance Criteria

```[tasklist]
### The items to complete to satisfy the Definition of Done.
- [x] All the unit tests have been executed and passed locally after PR work is complete. _(Why? Refer to 'Additional Information' below.)_
- [x] The file _**TextureFactoryTests.cs**_ has been refactored
- [x] The file _**TextureTests.cs**_ has been refactored
- [x] The file _**GpuDataTypeExtensionsTests.cs**_ has been refactored
- [x] The file _**NativelnputFactoryTests.cs**_ has been refactored
- [x] The file _**ImageLoaderTests.cs**_ has been refactored
- [x] The file _**RenderMediatorTests.cs**_ has been refactored
- [x] The file _**FontRendererTests.cs**_ has been refactored
- [x] The file _**LineRendererTests.cs**_ has been refactored
- [x] The file _**ShapeRendererTests.cs**_ has been refactored
- [x] The file _**TextureRendererTests.cs**_ has been refactored
- [x] The file _**SystemDisplayServiceTests.cs**_ has been refactored
```

### ToDo Items

```[tasklist]
### The items to complete to satisfy the Definition of Done.
- [x] Add a change type label to the issue.  Refer to the _**Change Type Labels**_ section below.
- [x] Add a priority label to the issue.  Refer to the _**Priority Type Labels**_ section below.
- [x] Link the issue to the correct milestone _(if applicable)_.
```

### Issue Dependencies

_No response_

### Related Work

_No response_

### Additional Information:

**_<details closed><summary>Unit Tests</summary>_**

Reasons for local unit test execution:
- Unit tests might pass locally but not in the CI environment during the status check process or vice-versa.
- Tests might pass on the developer's machine but not necessarily on the code reviewer's machine.
- If you notice that the test status check has passed but the test failed locally, please notify a project maintainer!

💡Warning💡
If the unit tests pass remotely and are not executed locally, this means we could be letting a bug slip into production.
Though bugs will always exist in some capacity, we should all do our part to help prevent them from happening.
</details>

**_<details closed><summary>Change Type Labels</summary>_**

| Change Type              | Label                     |
|:-------------------------|:--------------------------|
| Bug Fixes                | `🐛bug`                   |
| Breaking Changes         | `🧨breaking-changes`      |
| New Feature              | `✨new-feature`           |
| CICD Changes             | `♻️cicd`                  |
| Config Changes           | `⚙️config`                |
| Performance Improvements | `🏎️performance`           |
| Code Doc Changes         | `🗒️documentation/code`    |
| Product Doc Changes      | `📝documentation/product` |
</details>

**_<details closed><summary>Priority Type Labels</summary>_**

| Priority Type       | Label             |
|---------------------|-------------------|
| Low Priority        | `low priority`    |
| Medium Priority     | `medium priority` |
| High Priority       | `high priority`   |
</details>

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct.

### PR DESCRIPTION
**Pull request by sbundaz**
### **Description:**

To allow this pull request to be merged, please make sure that the following items of this pull request are complete.

This pull request closes #1006

✅The pull request head branch must be a feature branch with the syntax 'feature/<issue-num>-sync-testing'.  
✅The pull request base branch must be a 'main' or 'preview' branch..  
✅The pull request head branch contains a valid issue number.  
✅The pull request title matches the linked issue title exactly.  
✅The pull request assignees match the assignees of the issue.  
✅The pull request labels match the labels of the issue.  
✅The pull request organizational projects match the organizational projects of the issue.  
✅The pull requests milestone matches the milestone of the issue.  

---

### **Additional Info:**
1. Pull requests are automatically synced with the associated issue upon creation by one of the KinsonDigital workflow bots. 
2. The associated issue is the issue number that is embedded in the pull request head branch.
3. The list above will be automatically updated as the pull request's various settings match or do not match the associated issue.
4. To manually sync the pull request to the issue, create a comment with the `[run-sync]` command.
> [!Note] You must be an admin member of the organization to use this command.

---

- [x] Sync with the issue.  Use this to enable or disable this pull request from syncing with its associated issue.

### CODE REVIEW

File: Testing/VelaptorTests/Content/TextureTests.cs [hunk 1]
```diff
@@ -291,14 +293,16 @@ public void InternalCtor_WhenInvoked_UploadsTextureDataToGpu()
             0,
             GLPixelFormat.Rgba,
             GLPixelType.UnsignedByte,
-            expectedPixelArray), Times.Once());
+            Arg.Is<byte[]>(actualPixelArray => actualPixelArray.SequenceEqual(expectedPixelArray)));
```
Comment (sbundaz) [inline Testing/VelaptorTests/Content/TextureTests.cs:296]: @KinsonDigitalAdmin I'm not sure about this refactor. More info here: https://stackoverflow.com/questions/12699066/nsubstitute-checking-received-methods-with-array-arguments

Let me know if you have a better solution

Comment (CalvinWilkinson) [reply Testing/VelaptorTests/Content/TextureTests.cs:296]: This works fine.  NSubstitute basically sends the arg to the lambda method, which is, in your case, `actualPixelArray.`  There is another way by using `Callback` but this is more elegant is is how you should do this.

File: Testing/VelaptorTests/Graphics/Renderers/FontRendererTests.cs [hunk 2]
```diff
@@ -69,45 +70,53 @@ public class FontRendererTests : TestsBase
...
+        mockPushReactable
+            .When(x => x.Subscribe(Arg.Any<IReceiveSubscription>()))
+            .Do(ci =>
+            {
+                var reactor = ci.Arg<IReceiveSubscription>();
+                this.batchHasBegunReactor = reactor;
+            });
 
-        var mockFontRenderBatchReactable = new Mock<IRenderBatchReactable<FontGlyphBatchItem>>();
+        var mockFontRenderBatchReactable = Substitute.For<RenderBatchReactable<FontGlyphBatchItem>>();
```
Comment (CalvinWilkinson) [inline Testing/VelaptorTests/Graphics/Renderers/FontRendererTests.cs]: Please change to what is shown below:

```suggestion
        var mockFontRenderBatchReactable = Substitute.For<IRenderBatchReactable<FontGlyphBatchItem>>();
```

**Reason:**
When it comes to classes, NSubstitute (and most of mocking frameworks) cannot mock a class that is set to 'sealed'.  The class `RenderBatchReactable` is a sealed class so this is the reason for all of the tests failing for this test class.

Changing this to `IRenderBatchReactable` will fix this issue.

**Review by CalvinWilkinson — CHANGES_REQUESTED** [review]

File: Testing/VelaptorTests/Graphics/Renderers/FontRendererTests.cs (hunk 2 above)
Comment (sbundaz) [reply Testing/VelaptorTests/Graphics/Renderers/FontRendererTests.cs]: Done.

**Review by CalvinWilkinson — APPROVED** [review]

**Review by CalvinWilkinson — APPROVED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]

**Pull request merged**